modules/firebase/firebase.go

```diff
@@ -0,0 +1,165 @@
+package firebase
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"reflect"
+	"slices"
+
+	"github.com/docker/docker/api/types/mount"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+// Container represents the Firebase container type used in the module
+type Container struct {
+	testcontainers.Container
+}
+
+const rootFilePath = "/srv/firebase"
+
+// WithRoot sets the directory which is copied to the destination container as firebase root
+func WithRoot(rootPath string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) error {
+		req.Files = append(req.Files, testcontainers.ContainerFile{
+			HostFilePath:      rootPath,
+			ContainerFilePath: rootFilePath,
+			FileMode:          0o775,
+		})
+
+		return nil
+	}
+}
+
+// WithData names the data directory (by default under firebase root), can be used as a way of setting up fixtures.
+// Usage of absolute path will imply that the user knows how to mount external directory into the container.
+func WithData(dataPath string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) error {
+		req.Env["DATA_DIRECTORY"] = dataPath
+		return nil
+	}
+}
+
+const cacheFilePath = "/root/.cache/firebase"
+
+// WithCache enables firebase binary cache based on session (meaningful only when multiple tests are used)
+func WithCache() testcontainers.CustomizeRequestOption {
+	volumeName := fmt.Sprintf("firestore-cache-%s", testcontainers.SessionID())
+	volumeOptions := &mount.VolumeOptions{
+		Labels: testcontainers.GenericLabels(),
+	}
+
+	return func(req *testcontainers.GenericContainerRequest) error {
+		m := testcontainers.ContainerMount{
+			Source: testcontainers.DockerVolumeMountSource{
+				Name:          volumeName,
+				VolumeOptions: volumeOptions,
+			},
+			Target: cacheFilePath,
+		}
+		req.Mounts = append(req.Mounts, m)
+		return nil
+	}
+}
+
+func gatherPorts(config FirebaseConfig) ([]string, error) {
```

Comment:
question: could you describe what this is doing as it looks like its just getting the ports from UI field which wouldn't require reflection.

Reply:
It gathers the port numbers from config.emulators sub-block.

Instead of reflection here, I could demolish the structure definition of the `partialFirebaseConfig`, parse the json into `map[string]interface{}` and then extract matching paths by doing key comparisons (I need to get all values under the glob `config.emulators.*.port` and `config.emulators.*.websocketPort`).

However, I think that having the structure definition copied over from the typescript library is a bit more comprehensive, and if the emulator config schema changes, it will be easier to see what has changed from a structure definition.